pkg/reconciler/v1alpha1/autoscaling/hpa/hpa_test.go

```diff
@@ -36,6 +41,45 @@ const (
 	testRevision  = "test-revision"
 )
 
+func TestControllerCanReconcile(t *testing.T) {
+	kubeClient := fakeK8s.NewSimpleClientset()
+	servingClient := fakeKna.NewSimpleClientset()
+
+	opts := reconciler.Options{
+		KubeClientSet:    kubeClient,
+		ServingClientSet: servingClient,
+		Logger:           TestLogger(t),
+	}
+
+	servingInformer := informers.NewSharedInformerFactory(servingClient, 0)
+	kubeInformer := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+
+	ctl := NewController(&opts,
+		servingInformer.Autoscaling().V1alpha1().PodAutoscalers(),
+		kubeInformer.Autoscaling().V1().HorizontalPodAutoscalers(),
+	)
+
+	podAutoscaler := pa(testRevision, testNamespace, WithHPAClass)
+	servingClient.AutoscalingV1alpha1().PodAutoscalers(testNamespace).Create(podAutoscaler)
+	servingInformer.Autoscaling().V1alpha1().PodAutoscalers().Informer().GetIndexer().Add(podAutoscaler)
+
+	reconcileDone := make(chan struct{})
+	go func() {
+		defer close(reconcileDone)
+		err := ctl.Reconciler.Reconcile(context.TODO(), testNamespace+"/"+testRevision)
+		if err != nil {
+			t.Errorf("Reconcile() = %v", err)
```

Comment:
Oh interesting.  This pattern is from the revision controller some time ago.  It looks like they are using a WaitForHooks helper function now, probably to avoid this data race.

Reply:
The HPA controller doesn't actually need to reconcile in a separate goroutine.  So I've just pulled it out, which resolves the issue on this PR.

However I will open a separate issue to fix the error in the KPA controller: https://github.com/knative/serving/blob/9c525d989ff9a5c52907f02e92b429eda1fd5a77/pkg/reconciler/v1alpha1/autoscaling/kpa/kpa_test.go#L123  Thanks!

Reply:
FYI: the KPA controller (not this PR) needs to use multiple goroutines because the KPAMetrics in-memory resources needs to reconcile for the KPA-class PodAutoscaler reconcilation to complete.  A single goroutine cannot make progress.